### PR TITLE
feat(mysql): support max_execution_time optimizer hint

### PR DESCRIPTION
**Pull request by SeongJaeSong**
## Pull Request Checklist

- [X] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [X] Did you update the typescript typings accordingly (if applicable)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Relates issue
https://github.com/sequelize/sequelize/issues/15173

### CODE REVIEW

File: src/dialects/abstract/query-generator.js
Comment (fzn0x) [inline src/dialects/abstract/query-generator.js:2228]: Where is this number coming from?

Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js:2228]: It's in [MySQL Docs](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_max_execution_time).

**Review by fzn0x — APPROVED** [review]

File: src/model.d.ts
Comment (fzn0x) [inline src/model.d.ts:824]: I think this section could be more informative.

```js
/**
* This sets the max execution time for MySQL
*/
```

File: src/dialects/abstract/query-generator.js
Comment (fzn0x) [inline src/dialects/abstract/query-generator.js:2234]:
```suggestion
        throw new Error(`options.maxExecutionTimeMs value is must be between ${MINIMUM_EXECUTION_TIME_VALUE} and ${MAXIMUM_EXECUTION_TIME_VALUE}`);
```

File: src/dialects/abstract/query-generator.js
Comment (ephys) [inline src/dialects/abstract/query-generator.js:2226]: Our new policy is to always throw if someone uses an option that's not supported by their dialect:

```suggestion
    if (options.maxExecutionTimeMs != null) {
      if (!this.dialect.supports.maxExecutionTimeHint) {
        throw new Error(`The maxExecutionTimeMs option is not supported by ${this.dialect.name}`);
      }
```

File: src/dialects/abstract/query-generator.js
Comment (ephys) [inline src/dialects/abstract/query-generator.js:2230]: Let's throw if someone uses a number with decimals, NaN, or the Infinities

```suggestion
      if (Number.isSafeInteger(options.maxExecutionTimeMs) && options.maxExecutionTimeMs >= MINIMUM_EXECUTION_TIME_VALUE
```

File: src/dialects/abstract/query-generator.js
Comment (ephys) [inline src/dialects/abstract/query-generator.js:2234]: Inspect comes from the `node:util` package

```suggestion
        throw new Error(`maxExecutionTimeMs must be between ${MINIMUM_EXECUTION_TIME_VALUE} and ${MAXIMUM_EXECUTION_TIME_VALUE}, but it is ${inspect(options.maxExecutionTimeMs)}`);
```

File: src/dialects/mysql/index.ts
Comment (ephys) [inline src/dialects/mysql/index.ts:57]: does mariadb support this too?

File: test/unit/dialects/mysql/query-generator.test.js
Comment (ephys) [inline test/unit/dialects/mysql/query-generator.test.js:703]: This test file is (slowly) being phased out, we run all tests against all dialects now

Could you remove these tests and add them to https://github.com/sequelize/sequelize/blob/main/test/unit/query-generator/select-query.test.ts instead?

Some dialects will throw but that should be tested too. Here is how we do it for another queryGenerator method: https://github.com/sequelize/sequelize/blob/main/test/unit/query-generator/create-database-query.test.ts#L80-L85

**Review by ephys — CHANGES_REQUESTED** [review]
Thank you for taking the time to add a new feature to sequelize

I would call the option `maxExecutionTimeHintMs` instead of `maxExecutionTimeMs`. It's likely that we'll add a feature that enforces a max execution time that works in all dialects and I don't know yet how compatible it would be with the hint from mysql.

File: src/dialects/mysql/index.ts
Comment (SeongJaeSong) [reply src/dialects/mysql/index.ts:57]: MariaDB support functionally same thing([max_statement_time](https://mariadb.com/kb/en/aborting-statements/)), but it's a little bit diferrent.
1. unit time: `second` vs `millisecond`
2. name: `max_statement_time` vs `max_execution_time`
3. syntax: `SET STATEMENT max_statement_time=100 FOR SELECT ~` vs `SELECT /*+ MAX_EXECUTION_TIME(N) */ FROM ~`

Comment (ephys) [reply src/dialects/mysql/index.ts:57]: Ok then this PR can be exclusively for mysql

File: test/unit/dialects/mysql/query-generator.test.js
Comment (SeongJaeSong) [reply test/unit/dialects/mysql/query-generator.test.js]: I wrote the test code against all dialects(include throw when unsupported option.

In mssql case, they overwrite the abstract query generator's `selectFromTableFragment` method. So I added condition statement in mssql query generator's `selectFromTableFragment` method to make pass the test.

But I wonder is this correct way. Is there any option better this way?

File: src/dialects/mysql/index.ts
Comment (WikiRik) [reply src/dialects/mysql/index.ts:57]: Maybe we should implement this for `mariadb` as well in this PR? Even though it's `max_statement_time` I think that the same option would be fine to use. Possibly making this an object since mysql only supports this for `SELECT` queries but mariadb for different queries. (and then we'll just do the `SELECT` support in this PR to keep the scope limited)

File: src/dialects/abstract/query-generator.js
Comment (ephys) [reply src/dialects/abstract/query-generator.js:2233]: Could you add a link to it so it's explicit that this is about the mysql implem? (since it's in the abstract query generator)

File: test/unit/dialects/mysql/query-generator.test.js
Comment (ephys) [reply test/unit/dialects/mysql/query-generator.test.js]: That's a good solution :) We have a function called `rejectInvalidOptions` that we typically use for this but right now this is fine

File: src/dialects/mysql/index.ts
Comment (ephys) [reply src/dialects/mysql/index.ts:57]: I'm OK with the actual implementation being done in a separate PR, but in preparation for that, I think `selectFromTableFragment` should call a new protected method `this._getSelectOptimizerHints` (returns `''` by default) and move the actual logic inside of `MySqlQueryGenerator`. MariaDB will need to replace it to return `''` since they use a different syntax.

MariaDB supports it on all queries, but right now we're only considering it for `SELECT`. Maybe name it `selectMaxExecutionTimeHint`? We can always change the flag if we decide to support other statements. 

Or alternatively:

```ts
// supported on all queries
maxExecutionTimeHint: true

// unsupported
maxExecutionTimeHint: true

// supported on these specific queries
maxExecutionTimeHint: ['select'],
// or?
maxExecutionTimeHint: { select: true },
```

Comment (WikiRik) [reply src/dialects/mysql/index.ts:57]: Yeah, I was thinking about `maxExecutionTimeHint: { select: true },`

Comment (SeongJaeSong) [reply src/dialects/mysql/index.ts:57]: Then should we keep this PR for mysql and implement for mariadb in a separate PR?

Comment (WikiRik) [reply src/dialects/mysql/index.ts:57]: Yeah, some preparation for mariadb is done in this PR but the actual implementation in a separate PR.
So the `_getSelectOptimizerHints` that @ephys mentioned is good for this PR, just as updating `maxExecutionTimeHint` to accept an object `{ select: true }`

File: src/dialects/abstract/query-generator.js
Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js]: I move actual logic to mysql query generator in [here](https://github.com/sequelize/sequelize/pull/15341/commits/48b32fc5cb0a3acd73b22419590ab5bbca5cbd54), so I resolve this.

File: src/dialects/mysql/index.ts
Comment (SeongJaeSong) [reply src/dialects/mysql/index.ts]: I implemented `_getSelectOptimzerHints` returns 'SELECT` by default. Because in mysql, optimizer hint locate after of 'SELECT' but in mariadb locate before of 'SELECT'.

File: src/model.d.ts
Comment (WikiRik) [inline src/model.d.ts:826]: I think we should rename this to `maxExecutionTimeHint` as well so we can use the same one for MariaDB. Or would that cause too much confusion?

File: test/unit/query-generator/select-query.test.ts
Comment (WikiRik) [inline test/unit/query-generator/select-query.test.ts:481]:
```suggestion
        default: notSupportedError,
        mysql: 'SELECT /*+ MAX_EXECUTION_TIME(1000) */ `id` FROM `Users` AS `User`;',
```

File: src/model.d.ts
Comment (SeongJaeSong) [reply src/model.d.ts:826]: Yeah, I think it may cause confusion. How do you think about keep it `~Ms` and mult thousand when implement for MariaDb?
ex)
```
_getSelectOptimizerHints(options) {
    return `SET STATEMENT max_statement_time=${options.maxExecutionTimeHintMs * 1000} FOR SELECT`;
  }

```

Comment (ephys) [reply src/model.d.ts:826]: I would go with MS and divide by 1000 + round for mariadb, otherwise we loose the ability to be precise for mysql

File: src/dialects/abstract/query-generator.js
Comment (ephys) [inline src/dialects/abstract/query-generator.js:2236]: I would actually keep SELECT here, it's not part of the hint

```suggestion
    let fragment = 'SELECT ' + this._getSelectOptimizerHints(options);
```

`_getSelectOptimizerHints` can return `''` instead

Comment (ephys) [reply src/dialects/abstract/query-generator.js:2236]: I think once this is done, we're good to merge :)

Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js:2236]: In that case, we should branch in abstract query generator to make query for MySQL and MariaDB seperately like below.
```
let fragment = 'SELECT';
if (this.dialect.name === 'mysql') {
  fragment = fragment + this._getSelectOptimizerHints(options);
} else if (this.dialect.name === 'mariadb') {
  fragment = this._getSelectOptimizerHints(options) + fragment;
} else {
  ...
}
```
Because (I said in [above conversation](https://github.com/sequelize/sequelize/pull/15341#discussion_r1038752650)) MariaDb and MySQL are have difference to make optimizer hint - MySQL's optimizer hint must be located after SELECT(`SELECT /*+ MAX_EXECUTION_TIME(N) */ FROM ~`) but MariaDB's must before SELECT(`SET STATEMENT max_statement_time=100 FOR SELECT`).

The reason I'm reluctant to branch from abstract query generators is that I believe abstracts shouldn't be materialized. What do you think? @ephys

File: src/dialects/abstract/query-generator.js
Comment (ephys) [inline src/dialects/abstract/query-generator.js:2240]: Wait but actually we don't need this. That can already be done by overriding `selectFromTableFragment` to return something extra at the beginning of the string

Only `_getAfterSelectFragment` is necessary

File: src/dialects/abstract/query-generator.js
Comment (ephys) [inline src/dialects/abstract/query-generator.js:2242]: I still recommend naming this one something more explicit like `_getBeforeSelectAttributesFragment`. We could need to add more fragments in the future and the current name makes me think it goes at the end of the select query

**Review by ephys — CHANGES_REQUESTED** [review]
Thanks for your patience

File: src/dialects/abstract/query-generator.js
Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js:2240]: I don't understand about you said only `_getAfterSelectFragment` is necessary. I think if we will overriding `selectFromTableFragment` to return something extra at the beginning of the string, it can be return a something extra at the end of the string too.

In that context, Is `_getAfterSelectFragment` unnecessary too?

Comment (ephys) [reply src/dialects/abstract/query-generator.js:2240]: In the code you had, `_getAfterSelectFragment` wasn't added to the end of the string but inside of it

To be sure there isn't a misunderstanding: I'm talking about `_getBeforeSelectAttributesFragment`, which was called `_getAfterSelectFragment` before. A `_getAfterSelectFragment` that's added at t he end of the string would be unnecessary too

Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js:2240]: Ok, now I understand why `_getAfterSelectFragment(same with _getBeforeSelectAttributesFragment)` is necessary.

Then after we delete unnecessary `_getBeforeSelectFragment` function, how we support optimizer hint in MariaDB? By overriding `selectFromTableFragment`?

I think overriding `selectFromTableFragment` is take more cost than overriding `_getBeforeSelectFragment`. Because if we override `selectFromTableFragment`, we need to implement entire `selectFromTableFragment` logic.(I think `selectFromTableFragment` have more responsibility than `_getBeforeSelectFragment`)

But if we use `_getBeforeSelectFragment` method, we can support that by implement(override) `_getBeforeSelectFragment` only in MariaDB query generator.

Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js:2240]: Ok, now I understand why `_getAfterSelectFragment(same with _getBeforeSelectAttributesFragment)` is necessary.

Then after we delete unnecessary `_getBeforeSelectFragment` function, how we support optimizer hint in MariaDB? By overriding `selectFromTableFragment`?

I think overriding `selectFromTableFragment` is take more cost than overriding `_getBeforeSelectFragment`. Because if we override `selectFromTableFragment`, we need to implement entire `selectFromTableFragment` logic.(I think `selectFromTableFragment` have more responsibility than `_getBeforeSelectFragment`)

But if we use `_getBeforeSelectFragment` method, we can support that by implement(override) `_getBeforeSelectFragment` only in MariaDB query generator.

@ephys Thanks for advise.

Comment (ephys) [reply src/dialects/abstract/query-generator.js:2240]: > I think overriding selectFromTableFragment is take more cost than overriding _getBeforeSelectFragment. Because if we override selectFromTableFragment, we need to implement entire selectFromTableFragment logic.

not at all, just need to do this:

```ts
class MariaDbQueryGenerator extends MySqlQueryGenerator {
  selectFromTableFragment(options, model, attributes, tables, mainTableAs) {
    return 'SET ...' + super.selectFromTableFragment(options, model, attributes, tables, mainTableAs);
  }
}
```

Comment (SeongJaeSong) [reply src/dialects/abstract/query-generator.js:2240]: I was dumb.

**Comment by SeongJaeSong** [comment]
@ephys Is there a problem that we can't merge this?

**Comment by WikiRik** [comment]
Can you resolve the merge conflicts?

File: src/dialects/mssql/query-generator.js
Comment (WikiRik) [inline src/dialects/mssql/query-generator.js:884]: This whole if-statement, including the above constants of `dbVersion` and `isSQLServer2008` has been removed from the codebase. Can you also remove it from your changes?

**Review by WikiRik — DISMISSED** [review]
Looks good to me. I'll give @ephys some time to look at it as well

**Comment by SeongJaeSong** [comment]
@WikiRik Is there any problem with this PR? If there is no plan to merge, I will close it.

**Comment by WikiRik** [comment]
@SeongJaeSong thanks for the reminder! If you can resolve the merge conflicts I think we can merge this in

**Review by WikiRik — APPROVED** [review]

**Pull request merged**
